src/markov/markov_approx.jl

```diff
@@ -214,3 +218,402 @@ function estimate_mc_discrete{T}(X::Vector{T})
+    # Sometimes the algorithm fails to converge if the initial guess is too far
+    # away from the truth. If this occurs, the program tries an initial guess
+    # of all zeros.
+    if !Optim.converged(res) && all(lambda0 .!= 0.0)
+        warn("Failed to find a solution from provided initial guess. Trying new initial guess.")
+        res = Optim.optimize(obj, grad!, hess!, zeros(lambda0), Optim.Newton(), options)
+    end
+    # check convergence
+    Optim.converged(res) || error("Failed to find a solution.")
+    # Compute final probability weights and moment errors
+    lambdaBar = Optim.minimizer(res)
+    minimum_value = Optim.minimum(res)
+
+    Tdiff = Tx-repmat(TBar, 1, N)
+    p = (q.*exp.(lambdaBar'*Tdiff))/minimum_value
+    grad = similar(lambda0)
+    grad!(Optim.minimizer(res), grad)
+    momentError = grad/minimum_value
+    return p, lambdaBar, momentError
+end
+
+"""
+Compute the moment defining function used in discrete_approximation
+    ```julia
+    T = polynomial_moment(X,mu,scalingFactor,nMoment)
+    ```
+##### Inputs:
+- `X::Vector` : (N x 1) vector of grid points
+- `mu::Real` : location parameter (conditional mean)
+- `scalingFactor::Real` : scaling factor for numerical stability (typically largest grid point)
+- `nMoments::Integer` : number of polynomial moments
+##### Return
+- `T` : moment defining function used in discrete_approximation
+"""
+function polynomial_moment(X::Vector, mu::Real, scalingFactor::Real, nMoments::Integer)
+    # Check that scaling factor is positive
+    scalingFactor>0 || error("scalingFactor must be a positive number")
+
+    Y = (X-mu)/scalingFactor # standardized grid
+    T = Y'.^collect(1:nMoments)
+end
+
+"""
+Compute the maximum entropy objective function used in discrete_approximation
+    ```julia
+    obj = entropy_obj(lambda,Tx,TBar,q)
+    ```
+##### Arguments
+- `lambda::Vector` : (L x 1) vector of values of the dual problem variables
+- `Tx::Matrix` : (L x N) matrix of moments evaluated at the grid points
+         specified in discrete_approximation
+- `TBar::Vector` : (L x 1) vector of moments of the underlying distribution
+           which should be matched
+- `q::Matrix` : (1 X N) vector of prior weights for each point in the grid.
+
+##### Returns
+- `obj` : scalar value of objective function evaluated at lambda
+"""
+function entropy_obj(lambda::Vector, Tx::Matrix, TBar::Vector, q::Matrix)
+
+    # Some error checking
+    L, N = size(Tx)
+
+    !(length(lambda) != L || length(TBar) != L || length(q) != N) || error("Dimensions of inputs are not compatible.")
+
+    # Compute objective function
+    Tdiff = Tx-repmat(TBar, 1, N)
+    temp = q.*exp.(lambda'*Tdiff)
+    obj = sum(temp)
+    return obj
+end
+
+"""
+Compute gradient of objective function
+##### Returns
+- `grad` : (L x 1) gradient vector of the objective function evaluated
+           at lambda
+"""
+function entropy_grad!(lambda::Vector, grad::Vector, Tx::Matrix, TBar::Vector, q::Matrix)
```

Comment:
Julia convention is typically to put the mutated arguments first -- Could we make `grad` the first argument?